### ISSUE TITLE
Expressive 3: Error Response and Not Found Response not issolated for each $app->handle() 

### ISSUE THREAD

**Issue by wshafer** [zendframework/zend-expressive#560]
- [x] I was not able to find an [open](https://github.com/zendframework/zend-expressive/issues?q=is%3Aopen) or [closed](https://github.com/zendframework/zend-expressive/issues?q=is%3Aclosed) issue matching what I'm seeing.
 - [x] This is not a question. (Questions should be asked on [slack](https://zendframework.slack.com/) ([Signup for Slack here](https://zendframework-slack.herokuapp.com/)) or our [forums](https://discourse.zendframework.com/).)

Provide a narrative description of what you are trying to accomplish.

When using event loops like Swoole or ReactPHP each call to handle() that results in an error or not found response is adding to the existing open stream.

Since the error response and the not found response are retrieved through the container, a new response with a new stream is not created.  Instead the existing open stream is used for each call causing multiple requests to bleed together.

### Code to reproduce the issue

```php

$response = $app->handle($request);
$response = $app->handle($request);

print $response->getBody()->getContents();
```

### Expected results
Only one error or not found page in the response body.

### Actual results
getContents() includes multiple responses.

**Comment by xtreamwayz** [comment]
> Since the error response and the not found response are retrieved through the container, a new response with a new stream is not created. Instead the existing open stream is used for each call causing multiple requests to bleed together.

If you use zend-servicemanager as your container you can mark services as not sharable. This might solve your issue:

```php
use const Zend\Expressive\NOT_FOUND_RESPONSE;
use const Zend\Expressive\Router\IMPLICIT_HEAD_MIDDLEWARE_RESPONSE;
use const Zend\Expressive\Router\IMPLICIT_HEAD_MIDDLEWARE_STREAM_FACTORY;
use const Zend\Expressive\Router\IMPLICIT_OPTIONS_MIDDLEWARE_RESPONSE;
use const Zend\Expressive\Router\METHOD_NOT_ALLOWED_MIDDLEWARE_RESPONSE;

return [
    'dependencies' => [
        'shared' => [
            NOT_FOUND_RESPONSE                      => false,
            IMPLICIT_HEAD_MIDDLEWARE_RESPONSE       => false,
            IMPLICIT_HEAD_MIDDLEWARE_STREAM_FACTORY => false,
            IMPLICIT_OPTIONS_MIDDLEWARE_RESPONSE    => false,
            METHOD_NOT_ALLOWED_MIDDLEWARE_RESPONSE  => false,
        ],
    ],
];

**Comment by Xerkus** [comment]
I believe that request and response should not be managed by container directly, instead it should manage factory that could be used to create new response by consumers.

**Comment by xtreamwayz** [comment]
I've looked into creating a `ResponseFactoryFactory`. It's certainly a fix for this, but it's a lot of work and requires changes to zend-expressive-router as well.

/cc @mwop thought?

**Comment by xtreamwayz** [comment]
Hah... Looks like it's already in progress: zendframework/zend-expressive-router#51

**Comment by weierophinney** [comment]
Fixed with #561 and zendframework/zend-expressive-router#51.
